File: mobile/lib/device.js
// Executes a device action on the phone via Android intents / URL schemes.
// This is the real (Expo Go-safe) layer of "Control Your Device": open apps,
// dial, message, navigate, search. Full in-app screen tapping needs a native
// Accessibility build (phase 2).
import { Linking, Platform } from "react-native";
import * as IntentLauncher from "expo-intent-launcher";
import { launchApp as nativeLaunch, hasAgent } from "./agent";

// Package names — native launch by package is the most reliable "open app".
const PKG = {
  whatsapp: "com.whatsapp", youtube: "com.google.android.youtube",
  instagram: "com.instagram.android", maps: "com.google.android.apps.maps",
  chrome: "com.android.chrome", google: "com.google.android.googlequicksearchbox",
  gmail: "com.google.android.gm", spotify: "com.spotify.music",
  phonepe: "com.phonepe.app", paytm: "net.one97.paytm",
  gpay: "com.google.android.apps.nbu.paisa.user", facebook: "com.facebook.katana",
  telegram: "org.telegram.messenger", playstore: "com.android.vending",
};

const enc = encodeURIComponent;

// app -> primary scheme (with a web/store fallback where useful)
const APP_SCHEME = {
  whatsapp: "whatsapp://send",
  youtube: "vnd.youtube://",
  instagram: "instagram://app",
  maps: "geo:0,0",
  chrome: "googlechrome://",
  gmail: "googlegmail://",
  spotify: "spotify://",
  playstore: "market://",
  facebook: "fb://",
  telegram: "tg://",
  twitter: "twitter://",
  linkedin: "linkedin://",
  phone: "tel:",
};
const APP_WEB = {
  whatsapp: "https://web.whatsapp.com",
  youtube: "https://www.youtube.com",
  instagram: "https://instagram.com",
  maps: "https://maps.google.com",
  chrome: "https://www.google.com",
  gmail: "https://mail.google.com",
  spotify: "https://open.spotify.com",
  playstore: "https://play.google.com",
  facebook: "https://facebook.com",
  twitter: "https://twitter.com",
  linkedin: "https://linkedin.com",
};

async function open(url, fallback) {
  try {
    await Linking.openURL(url);
    return true;
  } catch (e) {
    if (fallback) {
      await Linking.openURL(fallback);
      return true;
    }
    throw e;
  }
}

// A short, human label for what an action will do (shown on the card button).
export function actionLabel(a) {
  switch (a?.kind) {
    case "open_app": return `Open ${cap(a.app)}`;
    case "call": return `Call ${a.number || ""}`.trim();
    case "sms": return `Message ${a.number || ""}`.trim();
    case "whatsapp": return `WhatsApp ${a.number || "chat"}`.trim();
    case "maps": return `Navigate: ${a.query || ""}`.trim();
    case "web": return `Search: ${a.query || a.url || ""}`.trim();
    case "youtube": return a.query ? `YouTube: ${a.query}` : "Open YouTube";
    case "email": return `Email ${a.to || ""}`.trim();
    default: return "Do it";
  }
}

// Outward actions we treat as needing an explicit OK (message/call/email).
export const NEEDS_OK = new Set(["call", "sms", "whatsapp", "email"]);

export async function executeDeviceAction(a) {
  if (!a || !a.kind) throw new Error("No action");
  switch (a.kind) {
    case "open_app":
      return openApp(a.app);
    case "call":
      return open(`tel:${clean(a.number)}`);
    case "sms": {
      const b = a.body ? `${Platform.OS === "ios" ? "&" : "?"}body=${enc(a.body)}` : "";
      return open(`sms:${clean(a.number)}${b}`);
    }
    case "whatsapp": {
      const phone = clean(a.number);
      const q = [phone ? `phone=${phone}` : "", a.message ? `text=${enc(a.message)}` : ""].filter(Boolean).join("&");
      return open(`whatsapp://send${q ? "?" + q : ""}`, `https://wa.me/${phone}${a.message ? `?text=${enc(a.message)}` : ""}`);
    }
    case "maps":
      return open(`geo:0,0?q=${enc(a.query || "")}`, `https://www.google.com/maps/search/?api=1&query=${enc(a.query || "")}`);
    case "web":
      return open(a.url || `https://www.google.com/search?q=${enc(a.query || "")}`);
    case "youtube":
      return open(a.query ? `https://www.youtube.com/results?search_query=${enc(a.query)}` : "https://www.youtube.com");
    case "email": {
      const q = [a.subject ? `subject=${enc(a.subject)}` : "", a.body ? `body=${enc(a.body)}` : ""].filter(Boolean).join("&");
      return open(`mailto:${a.to || ""}${q ? "?" + q : ""}`);
    }
    default:
      throw new Error(`Unknown action: ${a.kind}`);
  }
}

async function openApp(app) {
  // most reliable: native launch by package (custom build only)
  if (hasAgent && PKG[app]) {
    const ok = await nativeLaunch(PKG[app]);
    if (ok) return true;
  }
  if (Platform.OS === "android") {
    // system apps via intents
    if (app === "camera") return IntentLauncher.startActivityAsync("android.media.action.STILL_IMAGE_CAMERA");
    if (app === "settings") return Linking.openSettings();
    if (app === "photos") return IntentLauncher.startActivityAsync("android.intent.action.VIEW", { type: "image/*" });
    if (app === "files") return IntentLauncher.startActivityAsync("android.intent.action.GET_CONTENT", { type: "*/*" });
  }
  const scheme = APP_SCHEME[app];
  if (scheme) return open(scheme, APP_WEB[app]);
  if (APP_WEB[app]) return open(APP_WEB[app]);
  if (app === "google") return open("https://www.google.com");
  throw new Error(`I don't know how to open ${app} yet.`);
}

const clean = (n) => (n || "").replace(/[^\d+]/g, "");
const cap = (s) => (s ? s[0].toUpperCase() + s.slice(1) : s);

File: backend/nova/core/device.py
"""Turn a natural phone command into a structured device action.

The mobile app executes these on-device via Android intents / URL schemes
(open an app, dial, message, navigate, search). The heavy lifting (understanding
the request) happens here with the LLM; a rules fallback keeps it working when
Ollama is down.

Action shape (JSON):
  {"kind": "open_app", "app": "whatsapp"}
  {"kind": "call", "number": "+9198..."}
  {"kind": "sms", "number": "...", "body": "..."}
  {"kind": "whatsapp", "number": "...", "message": "..."}
  {"kind": "maps", "query": "..."}
  {"kind": "web", "query": "..."}          # or {"url": "https://..."}
  {"kind": "youtube", "query": "..."}
  {"kind": "email", "to": "...", "subject": "...", "body": "..."}
None  -> no executable action (Sarthi is just asking a question / chatting)
"""

from __future__ import annotations

import re
from typing import Optional, Tuple

# Apps the phone knows how to open (kept in sync with lib/device.js on mobile).
KNOWN_APPS = [
    "whatsapp", "youtube", "instagram", "maps", "chrome", "google", "camera",
    "settings", "phone", "gmail", "spotify", "photos", "files", "playstore",
    "facebook", "telegram", "twitter", "linkedin", "phonepe", "paytm", "gpay",
]

OUTWARD = {"call", "sms", "whatsapp", "email"}  # need approval before firing

_SYSTEM = (
    "You are Sarthi's phone-control planner. Convert the user's request into ONE "
    "device action as strict JSON. Respond with a JSON object: "
    '{"reply": <short spoken confirmation>, "action": <action object or null>}.\n'
    "Action kinds and fields:\n"
    '- open_app: {"kind":"open_app","app":"<one of: ' + "|".join(KNOWN_APPS) + '>"}\n'
    '- call: {"kind":"call","number":"<digits>"}\n'
    '- sms: {"kind":"sms","number":"<digits>","body":"<text>"}\n'
    '- whatsapp: {"kind":"whatsapp","number":"<digits or empty>","message":"<text>"}\n'
    '- maps: {"kind":"maps","query":"<place or address>"}\n'
    '- web: {"kind":"web","query":"<search text>"}\n'
    '- youtube: {"kind":"youtube","query":"<search text>"}\n'
    '- email: {"kind":"email","to":"<addr>","subject":"<s>","body":"<b>"}\n'
    "Rules: If a required detail is missing (e.g. a phone number you don't know, "
    "a contact name with no number), set action to null and ASK for it in reply. "
    "NEVER invent phone numbers, emails, or bracketed placeholders like [name]. "
    "Keep reply to one short sentence. Output JSON only."
)


_OPERATE_SYS = (
    "You are Sarthi operating an Android app for the user, one step at a time. "
    "You are given the GOAL and the list of on-screen elements (each has an index, "
    "text, and whether it's clickable/editable). Decide the SINGLE next action as "
    "JSON: {\"say\":\"<very short status>\",\"action\":\"tap|type|scroll|back|done|ask|confirm\","
    "\"text\":\"<tap target text OR text to type>\",\"label\":\"<for confirm/ask: what you're about to do>\"}.\n"
    "Rules: tap → text is the visible label to tap. type → first tap the input field in a "
    "previous step, then type. When the message/details are ready and only an irreversible "
    "SEND / PAY / CONFIRM tap remains, return action \"confirm\" with label describing it and "
    "text = the exact button label to tap (do NOT tap it yourself). Use \"done\" when the goal "
    "is achieved. Use \"ask\" if you're stuck or need info. Never invent contacts/amounts. JSON only."
)


def plan_operate_step(llm, goal: str, screen: list, history: list) -> dict:
    """Decide ONE in-app action from the current screen. Returns the action dict."""
    elems = "\n".join(
        f'[{i}] "{e.get("text","")}"'
        + (" (input)" if e.get("editable") else "")
        + (" (button)" if e.get("clickable") else "")
        for i, e in enumerate((screen or [])[:60])
    ) or "(screen is empty)"
    hist = "\n".join(f"- {h}" for h in (history or [])[-6:]) or "(none yet)"
    user = f"GOAL: {goal}\n\nDONE SO FAR:\n{hist}\n\nSCREEN:\n{elems}\n\nNext action as JSON:"
    try:
        if llm is not None and llm.available():
            out = llm.chat_json(_OPERATE_SYS, user)
            if isinstance(out, dict) and out.get("action"):
                return out
    except Exception:
        pass
    return {"action": "ask", "say": "I couldn't work out the next step.", "label": ""}


def plan_device_action(llm, text: str) -> Tuple[str, Optional[dict]]:
    text = (text or "").strip()
    if not text:
        return ("What would you like me to do on your phone?", None)

    # Try the model first.
    try:
        if llm is not None and llm.available():
            out = llm.chat_json(_SYSTEM, text)
            if isinstance(out, dict):
                reply = (out.get("reply") or "").strip() or "On it."
                action = out.get("action")
                action = _clean(action)
                return (reply, action)
    except Exception:
        pass

    # Rules fallback — covers the common asks without a model.
    return _rules(text)


def _clean(action):
    if not isinstance(action, dict):
        return None
    kind = (action.get("kind") or "").strip().lower()
    if kind not in {
        "open_app", "call", "sms", "whatsapp", "maps", "web", "youtube", "email"
    }:
        return None
    if kind == "open_app":
        app = (action.get("app") or "").strip().lower()
        if app not in KNOWN_APPS:
            return None
        return {"kind": "open_app", "app": app}
    # keep only string fields, drop empties we don't need
    out = {"kind": kind}
    for k, v in action.items():
        if k != "kind" and isinstance(v, (str, int)):
            out[k] = str(v)
    return out


def _rules(text: str) -> Tuple[str, Optional[dict]]:
    t = text.lower()

    # open an app
    m = re.search(r"\b(?:open|launch|khol|start)\s+([a-z ]+)", t)
    if m:
        name = m.group(1).strip().split()[0]
        alias = {"insta": "instagram", "wa": "whatsapp", "yt": "youtube", "map": "maps", "browser": "chrome"}
        name = alias.get(name, name)
        if name in KNOWN_APPS:
            return (f"Opening {name}.", {"kind": "open_app", "app": name})

    # navigation / maps
    m = re.search(r"\b(?:navigate to|directions to|maps? (?:to|for)|take me to)\s+(.+)", t)
    if m:
        q = m.group(1).strip()
        return (f"Opening maps for {q}.", {"kind": "maps", "query": q})

    # youtube / play
    m = re.search(r"\b(?:play|youtube|watch)\s+(.+)", t)
    if m:
        q = m.group(1).strip()
        return (f"Searching YouTube for {q}.", {"kind": "youtube", "query": q})

    # call
    m = re.search(r"\bcall\s+(.+)", t)
    if m:
        who = m.group(1).strip()
        num = re.sub(r"[^\d+]", "", who)
        if len(num) >= 7:
            return (f"Calling {num}.", {"kind": "call", "number": num})
        return (f"What's {who}'s number? I don't have it saved.", None)

    # whatsapp message
    m = re.search(r"\bwhatsapp\s+(.+)", t)
    if m:
        return ("Who should I message on WhatsApp, and what should it say?", None)

    # web search
    m = re.search(r"\b(?:search|google|find|look up)\s+(?:for\s+)?(.+)", t)
    if m:
        q = m.group(1).strip()
        return (f"Searching the web for {q}.", {"kind": "web", "query": q})

    return ("I can open apps, call, message on WhatsApp, navigate, or search — try 'open YouTube' or 'navigate to Connaught Place'.", None)
